modules/dbparser/pdb-file.c

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) 2002-2016 Balabit
+ * Copyright (c) 1998-2016 Balázs Scheidler
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "pdb-file.h"
+#include "pdb-error.h"
+#include "reloc.h"
+#include "pathutils.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+
+gint
+pdb_file_detect_version(const gchar *pdbfile, GError **error)
+{
+  FILE *pdb;
+  gchar line[1024];
+  gint result = 0;
+
+  pdb = fopen(pdbfile, "r");
+  if (!pdb)
+    {
+      g_set_error(error, PDB_ERROR, PDB_ERROR_FAILED, "Error opening file %s (%s)", pdbfile, g_strerror(errno));
+      return 0;
+    }
+
+  while (fgets(line, sizeof(line), pdb))
+    {
+      gchar *patterndb_tag;
+
+      patterndb_tag = strstr(line, "<patterndb");
+      if (patterndb_tag)
+        {
+          gchar *version, *start_quote, *end_quote;
+
+          /* ok, we do have the patterndb tag, look for the version attribute */
+          version = strstr(patterndb_tag, "version=");
+
+          if (!version)
+            goto exit;
+          start_quote = version + 8;
+          end_quote = strchr(start_quote + 1, *start_quote);
+          if (!end_quote)
+            {
+              goto exit;
+            }
+          *end_quote = 0;
+          result = strtoll(start_quote + 1, NULL, 0);
+          break;
+        }
+    }
+ exit:
+  fclose(pdb);
+  if (!result)
+    {
+      g_set_error(error, PDB_ERROR, PDB_ERROR_FAILED,
+                  "Error detecting pdbfile version, <patterndb> version attribute not found or <patterndb> is not on its own line");
+    }
+  return result;
+}
+
+static const gchar *
+_get_xsddir_in_build(void)
+{
+#ifdef SYSLOG_NG_ENABLE_DEBUG
+  const gchar *srcdir;
+  static gchar path[256];
+
+  srcdir = getenv("top_srcdir");
+  if (srcdir)
+    {
+      g_snprintf(path, sizeof(path), "%s/doc/xsd", srcdir);
```

Comment:
Hi,

The problem with allocating the string is that it makes calling this
function a lot more difficult, as the caller needs to free it.

This kind of pattern happens in syslog-ng a lot (e.g. formatting a string I
a static buffer and then returning that), and I would change them to use a
different pattern at once and not individually.

My point is that the caller shouldnt need to bother freeing the string as
that complicatedly call sites a lot. If we find a way how to do that, with
truncation and memory mgmt covered I am all for it.

BTW: truncation is not really a problem in this case but might in others.
On May 2, 2016 11:51 PM, "Thomas Nagy" <notifications@github.com> wrote:

> In modules/dbparser/pdb-file.c
> <https://github.com/balabit/syslog-ng/pull/1007#discussion_r61811173>:
>
> > +                  "Error detecting pdbfile version, <patterndb> version attribute not found or <patterndb> is not on its own line");
> > +    }
> > +  return result;
> > +}
> > +
> > +static const gchar *
> > +_get_xsddir_in_build(void)
> > +{
> > +#ifdef SYSLOG_NG_ENABLE_DEBUG
> > +  const gchar *srcdir;
> > +  static gchar path[256];
> > +
> > +  srcdir = getenv("top_srcdir");
> > +  if (srcdir)
> > +    {
> > +      g_snprintf(path, sizeof(path), "%s/doc/xsd", srcdir);
>
> We will need to check the return value of g_snprintf() as well. Would
> some kind of g_strdup_printf() work here instead?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/balabit/syslog-ng/pull/1007/files/9303b028cb03709baa83399624e236e7a808631c#r61811173>
>